pkg/openstack/instances.go

```diff
@@ -683,31 +692,19 @@ func getAddressesByName(client *gophercloud.ServiceClient, name types.NodeName,
 		return nil, err
 	}
 
-	interfaces, err := getAttachedInterfacesByID(client, srv.ID)
+	ports, err := getAttachedPorts(client, srv.ID)
 	if err != nil {
 		return nil, err
 	}
 
-	return nodeAddresses(&srv.Server, interfaces, networkingOpts)
+	return nodeAddresses(&srv.Server, ports, networkingOpts)
 }
 
-// getAttachedInterfacesByID returns the node interfaces of the specified instance.
-func getAttachedInterfacesByID(client *gophercloud.ServiceClient, serviceID string) ([]attachinterfaces.Interface, error) {
-	var interfaces []attachinterfaces.Interface
-
-	mc := metrics.NewMetricContext("server_os_interface", "list")
-	pager := attachinterfaces.List(client, serviceID)
-	err := pager.EachPage(func(page pagination.Page) (bool, error) {
-		s, err := attachinterfaces.ExtractInterfaces(page)
-		if err != nil {
-			return false, err
-		}
-		interfaces = append(interfaces, s...)
-		return true, nil
-	})
-	if mc.ObserveRequest(err) != nil {
-		return interfaces, err
+// getAttachedPorts returns a list of ports attached to a server.
+func getAttachedPorts(client *gophercloud.ServiceClient, serverID string) ([]ports.Port, error) {
+	listOpts := ports.ListOpts{
+		DeviceID: serverID,
```

Comment:
It's probably a minor use case, but I think we might want to filter by device_owner too even if Nova[^1] is arrogant enough to not consider other Neutron consumers. See https://bugs.launchpad.net/neutron/+bug/1982373 for context on how we hit that problem in the past. Even if it's "invalid" at the moment, I believe it should be fixed at some point.

[^1]: https://github.com/openstack/nova/blob/27.1.0/nova/api/openstack/compute/attach_interfaces.py#L74